scripts/changesummary.py

```diff
@@ -0,0 +1,540 @@
+            directory (str): path where the summary file should be saved
+        """
+
+        # Use a regex on the version to find versions with the pattern
+        # <v>.<0-9>.<0-9>.<0-9> . Any api that matches this pattern will be
+        # labeled as stable. In other words, v1, v1.4 and v1.4.5 is stable
+        # but v1b1 v1aplha and v1beta1 is not stable.
+        dataframe["IsStable"] = (
+            dataframe["Version"].str.extract(r"(v\d?\.?\d?\.?\d+$)").notnull()
+        )
+
+        # Create a filter for features, which contains only rows which have keys
+        # that have been deleted or added, that will be used as an argument in
+        # the `np.where()` call below.
+        filter_features = (dataframe["ChangeType"] == ChangeType.DELETED) | (
+            dataframe["ChangeType"] == ChangeType.ADDED
+        )
+
+        # Create a new column `IsFeature` to indicate which rows should be
+        # considered as features.
+        dataframe["IsFeature"] = np.where(filter_features, True, np.nan)
+
+        # Create a new column `IsFeatureAggregate` which will be used to
+        # summarize the api changes. We can either have feature or fix but not
+        # both.
+        dataframe["IsFeatureAggregate"] = dataframe.groupby("Name").IsFeature.transform(
+            lambda x: x.any()
+        )
+
+        # Create a new column `Summary`, which will contain a string with the
+        # conventional commit message.
+        dataframe["Summary"] = np.vectorize(self._build_summary_message)(
+            dataframe["Name"],
+            dataframe["IsFeatureAggregate"]
+        )
+
+        # Write the final dataframe to disk as it will be used in the
+        # buildprbody.py script
+        dataframe.to_csv(os.path.join(directory, "allapis.dataframe"))
+        return dataframe
+
+    def _write_verbose_changes_to_disk(self, dataframe, directory, summary_df):
+        """ "Writes verbose information to file about changes made to discovery
+        artifacts based on the provided dataframe. A separate file is saved
+        for each api in the current working directory. The extension of the
+        files will be `'.verbose'`.
+
+        args:
+            dataframe (object): a pandas dataframe containing verbose change
+                information for all discovery artifacts
+            directory (str): path where the summary file should be saved
+            summary_df (object): A dataframe containing a summary of the changes
+        """
+        # Array of strings which will contains verbose change information for
+        # each api
+        verbose_changes = []
+
+        # Select only the relevant columns. We need to create verbose output
+        # by Api Name, Version and ChangeType so we need to group by these
+        # columns.
+
+        change_type_groups = dataframe[
+            ["Name", "Version", "ChangeType", "Key", "Count"]
+        ].groupby(["Name", "Version", "ChangeType"])
```

Comment:
Could / should we also sort by name to minimize extra file operations?

Reply:
Done. https://github.com/googleapis/google-api-python-client/pull/1286/commits/00723563819ee741dc2dccd835fffde8dc8f479c